numbat/src/list.rs

```diff
@@ -0,0 +1,299 @@
+        }
+        // Second best case, the other slice comes from the same allocation and
+        // has the same view => they are equal
+        if Arc::ptr_eq(&self.alloc, &other.alloc) && self.view == other.view {
+            true
+        } else {
+            // Worst case scenario, we need to compare all the elements one by one
+            self.iter().zip(other.iter()).all(|(l, r)| l == r)
+        }
+    }
+}
+
+impl<T> NumbatList<T> {
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        if let Some(view) = self.view {
+            view.1 - view.0
+        } else {
+            self.alloc.len()
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            alloc: Arc::new(VecDeque::with_capacity(capacity)),
+            view: None,
+        }
+    }
+
+    /// Return the tail of the list without the first element.
+    /// Return an error if the list is empty.
+    pub fn tail(&mut self) -> Result<(), RuntimeError> {
+        if self.is_empty() {
+            return Err(RuntimeError::EmptyList);
+        }
+        if let Some(view) = &mut self.view {
+            view.0 += 1;
+            // should be ensured by the if above
+            debug_assert!(view.0 <= view.1);
+        } else {
+            self.view = Some((1, self.len()));
+        }
+        Ok(())
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        let (start, end) = self.view.map_or((0, self.alloc.len()), |view| view);
+        self.alloc.iter().skip(start).take(end - start)
+    }
+}
+
+impl<T: Clone> NumbatList<T> {
+    fn make_mut(&mut self) -> (&mut Option<(usize, usize)>, &mut VecDeque<T>) {
+        if Arc::strong_count(&self.alloc) != 1 {
+            // If someone else is using the list we must clone it
+            self.alloc = Arc::new(self.iter().cloned().collect());
+            self.view = None;
+        }
+        // With our usage, this should never allocate since we know we're the only
+        // one holding a reference to this `Arc` and we don't use weak references.
+        (&mut self.view, Arc::make_mut(&mut self.alloc))
+    }
+
+    /// Return the first element of the list. If we're the only owner of the list,
+    /// drop the list and do not copy anything. If another list is alive, only
+    /// clone the value that's being returned.
+    pub fn head(self) -> Option<T> {
+        let front = self.view.map_or(0, |(start, _end)| start);
+        if Arc::strong_count(&self.alloc) == 1 {
+            // safety: unwrap cannot fail because we ensured there was only one strong reference above
+            let mut alloc = Arc::try_unwrap(self.alloc).map_err(|_| ()).unwrap();
+            alloc.swap_remove_front(front)
+        } else {
+            self.alloc.get(front).cloned()
+        }
+    }
+
+    /// Allocate iif the list being used by another value at the same time
```

Comment:
Minor: Either just "if" or "iff" (if and only if). Similar below.
```suggestion
    /// Allocate if the list being used by another value at the same time
```